File: intro/c/calculator.c
#include <stdio.h>

// NOTE: data types - and initial values (optional)
int num_1;
int num_2;
char operator;
int result;

int sum(int x, int y) {
	int sum = x + y;
	return sum;
}

int calculate() {
	// NOTE: - calling predefined scanf function
	//	 - format string specifies data type; scanf function will use this
	//	 input parameter to know that it must cast our input ASCII string
	//	 into the binary representation of an integer value
	//	 - then, scanf uses the second input parameter to know where to store
	//	 the read integer value, here in the address of num_1
	// 	 - now, the value of num_1 will from 0 to our input, because the content
	//	 of the memory dedicated to num_1 was changed
	//	 - scanf is also defined in stdio.h like printf
	printf("Enter your first parameter:\n");
	scanf("%d", &num_1);
	printf("Enter your operator:\n");
	// NOTE: - format string specifies character data type
	// NOTE: - bug before adding newline to the format string
	//	   this is because a newline character is stuck in the input stream
	//	   from the entry of the input before this
	// NOTE: - debug this by looking at the output
	scanf("\n%c", &operator);
	printf("Enter your second parameter:\n");
	scanf("%d", &num_2);

	if (operator == '+') {
		result = sum(num_1, num_2);
	}

	// testing
	// NOTE: use of format string in C versus commas in python
	printf("%d %c %d = %d\n", num_1, operator, num_2, result);

	return result;
}

int main() {
	calculate();

	return 0;
}

/*
	GOTO build, see objdumps of data and bss and text regions
	GOTO gdb
	- explain debugging: pausing at certain times in execution and
	  observing the contents of the binary code - i.e. variables
	- layout split: - observe C code alongside the binary code
			- explain disassembly from binary signal to assembly text
	- print var, print/t var
	- x/b memory, x/x memory
	- note how the same binary pattern can be interpreted in different
	  ways to reflect different data types
*/

File: intro/python/calculator.py
## Learning Objectives:
#
#
#
# User interaction: reading input from user
# Conditional control flow operators: if/else


##############
num_1 = 0
num_2 = 0
operator = ''
result = 0

def add(x, y):
	sum = x + y
	return sum

def calculate():
	print("Enter your first parameter:")
	# NOTE: calling predefined input function
	num_1 = input()
	print("Enter your operator:")
	operator = input()
	print("Enter your second parameter:")
	num_2 = input()

	# NOTE: using if/else conditional construct
	# NOTE: using conditional operator ==
	# NOTE: conditional operators return a value, True or False,
	#	depending on how the expression evaluates
	if (operator == '+'):
		result = add(num_1, num_2)
	
	## testing
	print(num_1, operator, num_2, " = ", result)

#calculate()


### run and observe error as num_1 and num_2 are treated as ASCII strings 
### they were just concatenated as we have seen before

### the importance of data types
# though we are not forced to define them in python as we are in C,
# it is important to know the type of data you are dealing with
# in order for your code to run correctly

### here, as a user, i can only input ASCII text that i understand to the
# running program; the program itself should convert my ASCII text to 
# numeric data
# luckily, functions that do this already exist, and we can use them
# these functions are called casting functions; they convert the binary encoding
# of their parameters from one format to another.
## NOTE remember in the end of C we saw hexadecimal rep of binary encoding of 
##	both ascii strings and integers

def calculate():
	print("Enter your first parameter:")
	num_1_str = input()
	# NOTE casting/type conversion
	num_1 = int(num_1_str)
	print("Enter your operator:")
	operator = input()
	print("Enter your second parameter:")
	num_2_str = input()
	# NOTE casting/type conversion
	num_2 = int(num_2_str)

	if (operator == '+'):
		result = add(num_1, num_2)
	
	## testing
	print(num_1, operator, num_2, " = ", result)

#calculate()


## go to C and see binary, hex, integer, and character representations in GDB
## continue calculate function

def subtract(x, y):
	diff = x - y
	return diff

def multiply(x, y):
	product = x * y
	return product

def divide(x, y):
	quotient = x / y
	return quotient


def calculate():
	print("Enter your first parameter:")
	num_1_str = input()
	num_1 = int(num_1_str)
	print("Enter your operator:")
	operator = input()
	print("Enter your second parameter:")
	num_2_str = input()
	num_2 = int(num_2_str)

	# NOTE: nesting if/else structure
	#	understand nesting syntactically and semantically
	if (operator == '+'):
		result = add(num_1, num_2)
	else: 
		if (operator == '-'):
			result = subtract(num_1, num_2)
		else: 
			if (operator == '*'):
				result = multiply(num_1, num_2)
			else:
				if (operator == '/'):
					result = divide(num_1, num_2)

	## testing
	print(num_1, operator, num_2, " = ", result)

calculate()


### all runs well
### error handling?





